rust/datafusion/src/physical_plan/udf.rs

```diff
@@ -56,6 +56,12 @@ impl Debug for ScalarUDF {
     }
 }
 
+impl PartialEq for ScalarUDF {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name && self.signature == other.signature
```

Comment:
I am afraid that this may not be sufficient: two UDFs may have the same name and signature and represent different semantics. In general UDFs are anonymous functions and thus do not have a `partialEq`. If we base equality on `name` and two UDFs are assigned the same name, there will be a semantic error in this equality.

Note that while this can't happen in SQL, because UDFs are registered via `register_udf(name, ...)`, the DataFrame API supports using bypassing the context and calling UDFs without registering them on the context.

In the `simple_udf.rs` example, the line 

```rust
let expr1 = pow.call(vec![col("a"), col("b")]);
```

demonstrates how to use a UDF without registering it on the context, in which case its name is only used when printing the plan.

Reply:
Thanks for pointing this out, I was afraid there might be an issue like this. 

I'm happy to back out the `PartialEq` derivation for `Expr` (as well as `ScalarUDF` and `AggregateUDF`) , and do the equivalency checks manually in SQL planner as needed.
 
Before I do that, I'm wondering if you think there is a more suitable implementation of `PartialEq` for `ScalarUDF` and `AggregateUDF`? Maybe their `eq()` implementation can:

1) Compare the function pointers?
2) Always return `false`? 

I would advocate for some definition of equality for `Expr`, as, for example, I imagine an optimizer (outside the context of the SQL planner) should leverage equivalency of duplicated expressions. Of course equivalency doesn't need to be via `PartialEq`, but that would be a nice convenience. 

Let me know what you think. Thanks!

Reply:
I do not have a good answer here. Both options that you presented are reasonable. IMO we should just proceed with one, but let's double-check with @andygrove and @alamb ^_^

Reply:
I agree that we are performing comparison of expressions, we should have `partialEq` and not a generic function, as otherwise we may end up with multiple variations of a `eq`, which is error-prone.

Reply:
I think to handle it in the general case we would have to require that the user defined function / aggregate itself define equality (perhaps with a default implementation that compares function pointers as suggested by @drusso ).

I think user defined functions in this kind of framework are also tagged with other properties (like if they have side effects, and thus can't be optimized away) 

I personally suggest filing a ticket for this issue in the future -- it is kind of like the category of "more mature user defined function support".

Reply:
I've filed [ARROW-10963](https://issues.apache.org/jira/browse/ARROW-10963).